File: android/src/main/kotlin/ch/waio/pro_video_editor/ProVideoEditorPlugin.kt
package ch.waio.pro_video_editor

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import ch.waio.pro_video_editor.src.ExportVideo
import ch.waio.pro_video_editor.src.VideoProcessor
import ch.waio.pro_video_editor.src.ThumbnailGenerator
import kotlinx.coroutines.*
import java.io.File

/** ProVideoEditorPlugin */
class ProVideoEditorPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private lateinit var exportVideo: ExportVideo
    private lateinit var videoProcessor: VideoProcessor
    private lateinit var thumbnailGenerator: ThumbnailGenerator

    private val coroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "pro_video_editor")
        eventChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, "pro_video_editor_progress")

        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })

        exportVideo = ExportVideo(flutterPluginBinding.applicationContext);
        videoProcessor = VideoProcessor(flutterPluginBinding.applicationContext)
        thumbnailGenerator = ThumbnailGenerator(flutterPluginBinding.applicationContext)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "getVideoInformation" -> {
                val videoBytes = call.argument<ByteArray>("videoBytes")
                val extension = call.argument<String>("extension")

                if (videoBytes != null && extension != null) {
                    val info = videoProcessor.processVideo(videoBytes, extension)
                    result.success(info)
                } else {
                    result.error(
                        "InvalidArgument", "Expected raw Uint8List (ByteArray/List<Int>)", null
                    )
                }
            }

            "createVideoThumbnails" -> {
                val videoBytes = call.argument<ByteArray>("videoBytes")
                val rawTimestamps = call.argument<List<Number>>("timestamps")
                val thumbnailFormat = call.argument<String>("thumbnailFormat")
                val imageWidth = call.argument<Number>("imageWidth")?.toInt()
                val extension = call.argument<String>("extension")

                val timestamps = rawTimestamps?.map { it.toLong() }

                if (videoBytes == null || timestamps == null || thumbnailFormat == null || imageWidth == null || extension == null) {
                    result.error("INVALID_ARGUMENTS", "Missing or invalid arguments", null)
                    return
                }

                coroutineScope.launch {
                    try {
                        val thumbnails = thumbnailGenerator.generateThumbnails(
                            videoBytes = videoBytes,
                            timestamps = timestamps,
                            extension = extension,
                            thumbnailFormat = thumbnailFormat,
                            width = imageWidth
                        )

                        withContext(Dispatchers.Main) {
                            result.success(thumbnails)
                        }
                    } catch (e: Exception) {
                        withContext(Dispatchers.Main) {
                            result.error("THUMBNAIL_ERROR", e.message, null)
                        }
                    }
                }
            }

            "exportVideo" -> {
                val videoBytes = call.argument<ByteArray>("videoBytes")
                val imageBytes = call.argument<ByteArray>("imageBytes")
                val outputFormat = call.argument<String>("outputFormat") ?: "mp4"
                val preset = call.argument<String>("encodingPreset")
                val videoDuration = call.argument<Int>("videoDuration")
                val startTime = call.argument<Int>("startTime")
                val endTime = call.argument<Int>("endTime")
                val constantRateFactor = call.argument<Int>("constantRateFactor")
                val filters = call.argument<String>("filters") ?: ""
                val colorMatrices = call.argument<List<List<Double>>>("colorMatrices")
                
                if (videoBytes == null || imageBytes == null || videoDuration == null || 
                    preset == null || constantRateFactor == null) {
                    result.error(
                        "INVALID_ARGUMENTS",
                        "Missing parameters",
                        null
                    )
                    return
                }

                exportVideo.generate(videoBytes = videoBytes,
                    imageBytes = imageBytes,
                    outputFormat = outputFormat,
                    preset = preset,
                    startTime = startTime,
                    endTime = endTime,
                    videoDuration = videoDuration,
                    constantRateFactor = constantRateFactor,
                    filters = filters,
                    colorMatrices = colorMatrices,
                    onSuccess = { outputPath ->
                        val outputFile = File(outputPath)
                        val outputBytes = outputFile.readBytes()
                        Handler(Looper.getMainLooper()).post {
                            result.success(outputBytes)
                        }
                    },
                    onError = { errorMsg ->
                        Handler(Looper.getMainLooper()).post {
                            result.error("FFMPEG_ERROR", errorMsg, null)
                        }
                    },
                    onProgress = { progress ->
                        Handler(Looper.getMainLooper()).post {
                            eventSink?.success(progress)
                        }
                    })
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
        coroutineScope.cancel()
    }
}

File: ios/Classes/ProVideoEditorPlugin.swift
import Flutter
import UIKit

public class ProVideoEditorPlugin: NSObject, FlutterPlugin {
  var eventSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(name: "pro_video_editor", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "pro_video_editor_progress", binaryMessenger: registrar.messenger())
        
    let instance = ProVideoEditorPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)

    case "getVideoInformation":
      guard let args = call.arguments as? [String: Any],
            let videoData = args["videoBytes"] as? FlutterStandardTypedData,
            let ext = args["extension"] as? String else {
        result(["error": "Invalid arguments"])
        return
      }

      let output = VideoProcessor.processVideo(videoData: videoData.data, ext: ext)
      result(output)

    case "createVideoThumbnails":
      guard let args = call.arguments as? [String: Any],
            let videoData = args["videoBytes"] as? FlutterStandardTypedData,
            let rawTimestamps = args["timestamps"] as? [Double],
            let format = args["thumbnailFormat"] as? String,
            let ext = args["extension"] as? String,
            let width = args["imageWidth"] as? NSNumber else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing or invalid arguments", details: nil))
        return
      }

      ThumbnailGenerator.generateThumbnails(
        videoBytes: videoData.data,
        timestamps: rawTimestamps,
        thumbnailFormat: format,
        extension: ext,
        width: width
      ) { thumbnails in
        let flutterDataList = thumbnails.map { FlutterStandardTypedData(bytes: $0) }
        result(flutterDataList)
      }

     case "exportVideo":
        guard let args = call.arguments as? [String: Any],
              let videoData = args["videoBytes"] as? FlutterStandardTypedData,
              let imageData = args["imageBytes"] as? FlutterStandardTypedData,
              let outputFormat = args["outputFormat"] as? String,
              let preset = args["encodingPreset"] as? String,
              let videoDuration = args["videoDuration"] as? Int,
              let crf = args["constantRateFactor"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing parameters", details: nil))
            return
        }

        let startTime = args["startTime"] as? Int
        let endTime = args["endTime"] as? Int
        let filters = args["filters"] as? String ?? ""

        ExportVideo.generate(
            videoBytes: videoData.data,
            imageBytes: imageData.data,
            outputFormat: outputFormat,
            preset: preset,
            startTime: startTime,
            endTime: endTime,
            videoDuration: videoDuration,
            constantRateFactor: crf,
            filters: filters,
            onSuccess: { outputPath in
                if let fileData = try? Data(contentsOf: URL(fileURLWithPath: outputPath)) {
                    result(FlutterStandardTypedData(bytes: fileData))
                } else {
                    result(FlutterError(code: "FILE_ERROR", message: "Failed to read output file", details: nil))
                }
            },
            onError: { errorMessage in
                result(FlutterError(code: "FFMPEG_ERROR", message: errorMessage, details: nil))
            },
            onProgress: { progress in
                self.eventSink?(progress)
            }
        )

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}